### DISCUSSION TITLE
access Team model attribute like name by the roles() relationship

### DISCUSSION THREAD

**Discussion by samuel-ginot** [spatie/laravel-permission#2172] [Q&A] [answered]
Hello,

I have a project with Laravel Permission Package.
I have activate Teams Permission. All its ok.

But i don't know how I can access to my Team attributes like the name.

Should I create un belongsTo relation beetween roles() relation and Team Model?

In other words : if I have a team_id = 6 in my 'model_has_roles' table, how can I access to the Team name with id = 6.

thanks in advance
Sam

**Comment by erikn69** [comment, marked as answer]
> Should I create un belongsTo relation beetween roles() relation and Team Model?

If you need that, Yes, 
I have a belongsTo relation between the User and Team model, and I only needed that

>In other words : if I have a team_id = 6 in my 'model_has_roles' table, how can I access to the Team name with id = 6.

Also, you could try with `join` on that relation

**Reply by samuel-ginot** [reply]
Thanks a lot but I have surcharge this relation in my UserModel : 
`

public function roles(): BelongsToMany
    {

        $relation = $this->morphToMany(
            config('permission.models.role'),
            'model',
            config('permission.table_names.model_has_roles'),
            config('permission.column_names.model_morph_key'),
            PermissionRegistrar::$pivotRole
        );

        //$relation->belongsTo(Structure::class, 'structure_id');

        if (! PermissionRegistrar::$teams) {
            return $relation->withTimestamps();
        }

        return $relation->wherePivot(PermissionRegistrar::$teamsKey, getPermissionsTeamId())
            ->where(function ($q) {
                $teamField = config('permission.table_names.roles').'.'.PermissionRegistrar::$teamsKey;
                $q->whereNull($teamField)->orWhere($teamField, getPermissionsTeamId());
            })->withTimestamps();
    }
`

So how could i create a belongsTo relation with roles () relations?

**Reply by samuel-ginot** [reply]
SOrry this code : 

`public function roles(): BelongsToMany
    {

        $relation = $this->morphToMany(
            config('permission.models.role'),
            'model',
            config('permission.table_names.model_has_roles'),
            config('permission.column_names.model_morph_key'),
            PermissionRegistrar::$pivotRole
        );

        if (! PermissionRegistrar::$teams) {
            return $relation->withTimestamps();
        }

        return $relation->wherePivot(PermissionRegistrar::$teamsKey, getPermissionsTeamId())
            ->where(function ($q) {
                $teamField = config('permission.table_names.roles').'.'.PermissionRegistrar::$teamsKey;
                $q->whereNull($teamField)->orWhere($teamField, getPermissionsTeamId());
            })->withTimestamps();
    }`

**Reply by erikn69** [reply]
Please use correctly markdown ` ```php `, it is hard to read
 ```php
public function roles(): BelongsToMany
{
        $relation = $this->morphToMany(
        config('permission.models.role'),
        'model',
        config('permission.table_names.model_has_roles'),
        config('permission.column_names.model_morph_key'),
        PermissionRegistrar::$pivotRole
    );

    if (! PermissionRegistrar::$teams) {
        return $relation->withTimestamps();
    }

    // here try join with teams, read documentation for correct usage
    $relation->leftJoin('teams', 'id1', 'id2')->addSelect(\DB::raw('teams.name as team_name'));

    return $relation->wherePivot(PermissionRegistrar::$teamsKey, getPermissionsTeamId())
        ->where(function ($q) {
            $teamField = config('permission.table_names.roles').'.'.PermissionRegistrar::$teamsKey;
            $q->whereNull($teamField)->orWhere($teamField, getPermissionsTeamId());
        })->withTimestamps();
}
 ```

**Reply by erikn69** [reply]
> I've lost the others role fields...

Of course, it was just a guide, you must adjust the rest of the query, like adding `config('permission.models.role').'*'`

```php
$relation
  ->leftJoin('structures', 'structures.id', '=', 'model_has_roles.structure_id')
  ->select(
     'roles.*',
     // 'model_has_roles.*',
     \DB::raw('structures.name as structure_name')
  );
```

**Comment by samuel-ginot** [comment]
Sorry for markdown...
thanks it's a really good option ! But Like this I have only the team_name attribut . I've lost the others role fields...

for information I do This : 
```php
$relation = $this->morphToMany(
            config('permission.models.role'),
            'model',
            config('permission.table_names.model_has_roles'),
            config('permission.column_names.model_morph_key'),
            PermissionRegistrar::$pivotRole
        );

$relation
  ->join('structures', 'structures.id', '=', 'model_has_roles.structure_id')
  ->addSelect(\DB::raw('structures.name as structure_name'));
```

And Secondly, how can I retrieve the others structure table fields ?
Thanks
